### ISSUE TITLE
cmake hangs under xonsh; runs fine under bash

### ISSUE THREAD

**Issue by pletnes** [xonsh/xonsh#3020]
## xonfig

<details>
```
$ xonfig
+------------------+-----------------+
| xonsh            | 0.8.10          |
| Python           | 3.7.1           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | 2.0.8           |
| shell type       | prompt_toolkit2 |
| pygments         | 2.3.1           |
| on posix         | True            |
| on linux         | True            |
| distro           | unknown         |
| on darwin        | False           |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
+------------------+-----------------+
```
</details>

## Expected Behavior

I typed `cmake` and expected `cmake` to configure my project. 
## Current Behavior

I expected `cmake` to run just like when under `bash`. But... nothing happened. Turns out `cmake` hangs very early on. Even with debug output, just a single line of output appears. After quite a bit of fiddling, I swapped over to `bash` (gasp) and hit `cmake` - works like a charm.

Running `strace`, it appears that `cmake` hangs on a `select(...)` system call. The machine is ubuntu 16.04 under VMWare Workstation Pro. I've tried inside and outside of tmux, via ssh connections, and via the VMWare desktop GUI. Really, really, really weird...

The VM is quite new (apt and conda installed packages and a few config files) so I don't think there's any super weird config involved. I've tried removing `~/.xonshrc` but no luck there either.

### Traceback (if applicable)

<details>
```
traceback
```
</details>

## Steps to Reproduce

```
cmake ../path/to/CMakeLists.txt
```

I was trying to build pFUnit from here: https://github.com/Goddard-Fortran-Ecosystem/pFUnit

**Comment by scopatz** [comment]
Thanks for reporting @pletnes! I'll take a look.  Out of curiosity, does this work on xonsh v0.8.9?  We made some `close_fds` changes that may have affected CMake (they were made to fix make, ironically)

**Comment by pletnes** [comment]
Can I get xonsh0.8.9 from conda in any way?

**Comment by scopatz** [comment]
Yeah, I believe that `conda install -c conda-forge xonsh=0.8.9` will work

**Comment by gforsyth** [comment]
> We made some `close_fds` changes

I think it's impacted a few things -- I wasn't connecting the dots for a bit, but I've recently had to add `rm` to `threadable_predictors` locally because I use `rm -i` to confirm removals and the prompt gets closed before I can interact with it

**Comment by scopatz** [comment]
Hmm yeah, I could tell that change would be problematic

**Comment by pletnes** [comment]
Yeah, ok, so I tried your install instructions @scopatz  (btw I had to go back to bash to get conda to activate the env and do the installation there :-/) and tried running cmake. However, the behavior seems identical to me. I checked both `xonsh --version` and `xonfig` to ensure that I'd indeed opened a new shell process.

**Comment by scopatz** [comment]
Ok, that's good! That means it likely isn't some silly `close_fds` issue

**Comment by pletnes** [comment]
True, but doesn't help me build my Fortran code. Also I ran into https://github.com/conda/conda/issues/8274 on the way; unsure if this is conda's or xonsh' responsibility.

**Comment by scopatz** [comment]
I'll get to this ASAP

**Comment by pletnes** [comment]
Impressive response @scopatz! Also I just started to use xonsh a couple of days ago, and I'm blown away... Love it so far!

**Comment by scopatz** [comment]
I am still looking into this. I thought #3026 would fix it, but it doesn't :cry:

**Comment by pletnes** [comment]
Anything I can do to help? Did you reproduce the bug?

**Comment by scopatz** [comment]
Yeah I can reproduce it

**Comment by scopatz** [comment]
OK, I have figured out part of what is going on here, I think.  This is because of the fix in #3003.  I believe it is because we are setting a `SIG_BLOCK`, but then not setting a `SIG_UNBLOCK`.  I haven't found the right options that make everything work

**Comment by scopatz** [comment]
This was fixed by #3031
